---
title: "Fst"
author: "Vita"
date: "May 12, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(ggplot2)
library(gridExtra)
```



```{r import_data}
## paths
input <- "../input/"
output <- "../output/"
pic_path <- "../graphs/"

## import genotype data
## in letters
dAF <- fread(paste0(input, "genotypes_AF"), header = FALSE)
dEA <- fread(paste0(input, "genotypes_EA"), header = FALSE)
dSA <- fread(paste0(input, "genotypes_SA"), header = FALSE)
dWE <- fread(paste0(input, "genotypes_WE"), header = FALSE)
dAM <- fread(paste0(input, "genotypes_AM"), header = FALSE)
dO <- fread(paste0(input, "genotypes_O"), header = FALSE)
dCAS <- fread(paste0(input, "genotypes_CAS"), header = FALSE)

## as numeric matrix (for LD)
mAF <- readRDS(file = paste0(output, "AF", "_MT_Genotypes.RDS"))
mEA <- readRDS(file = paste0(output, "EA", "_MT_Genotypes.RDS"))
mSA <- readRDS(file = paste0(output, "SA", "_MT_Genotypes.RDS"))
mWE <- readRDS(file = paste0(output, "WE", "_MT_Genotypes.RDS"))
mAM <- readRDS(file = paste0(output, "AM", "_MT_Genotypes.RDS"))
mO <- readRDS(file = paste0(output, "O", "_MT_Genotypes.RDS"))
mCAS <- readRDS(file = paste0(output, "CAS", "_MT_Genotypes.RDS"))

## Snp data
dSnp <- fread(paste0(input, "snps_filtered")) 
colnames(dSnp) <- c("ID", "chrom", "pos", "base", "alt")
```

```{r calc_allele_freq}

# function to calculate NA in snp
get_NA_perc_forSNP <- function(Data, ...) {
  nNA_snp <- apply(Data, 1, function(x) sum(is.na(x)))
  max_nHap <- dim(Data)[2]
  df_NA_stat <- data.table(ID = 1:length(nNA_snp), 
                           n_na = nNA_snp,
                           perc_NA = nNA_snp / max_nHap) 
  #bin_width <- 100   # 100 SNP
  #df_NA_stat[, plot_bin := (as.integer(ID / bin_width) + 0.5) * bin_width]  
  return(df_NA_stat)
}

snp_na_perc <- get_NA_perc_forSNP(dAF)
summary(snp_na_perc)

## function to calculate allele frequency for each SNP from haplotype data 
calc_haplotype_freq <- function(dSnp, dGenotypes, ...) {
  
  nHap <- dim(dGenotypes)[2]
  dComb <- cbind(dSnp[, 3:5], dGenotypes)
  # count alternative allele
  n_A <- apply(dComb, 1, function(x) {
    a <- x[3]
    n_A <- sum(x[-c(1:3)] == a)
    })
  # count base allele
  n_B <- apply(dComb, 1, function(x) {
    b <- x[2]
    n_B <- sum(x[-c(1:3)] == b)
    })
  # combine results
  dRes <- data.table(pos = dComb$pos, 
                     f_A = n_A / nHap, 
                     f_B = n_B / nHap)
  # check if all frequencies add up to 1
  if (unique(dRes$f_A + dRes$f_B) != 1) {stop("allele frequencies do not add up to 1")}
  return(dRes)
}


f_AF <- calc_haplotype_freq(dSnp, dAF)
f_EA <- calc_haplotype_freq(dSnp, dEA)
f_SA <- calc_haplotype_freq(dSnp, dSA)
f_WE <- calc_haplotype_freq(dSnp, dWE)
f_O <- calc_haplotype_freq(dSnp, dO)
f_CAS <- calc_haplotype_freq(dSnp, dCAS)
f_AM <- calc_haplotype_freq(dSnp, dAM)
head(f_AM)

```


```{r way2:FST_between_pairs_of_populations}

## NOTE:some SNP are fixed in all populations that we are analysing. There is no variance at all in the pooled population: HT = 0. And then Fst becomes NA. Not interesting, so will remove those. However, I will not remove SNP who get Fst NA in just 2 populations. Could deduce then in which populations some SNP are fixated and in others not. 

dComb <- rbind(f_AF[, region := "AF"], 
               f_EA[, region := "EA"],
               f_SA[, region := "SA"],
               f_WE[, region := "WE"],
               f_O[, region := "O"],
               f_CAS[, region := "CAS"],
               f_AM[, region := "AM"])

dRes <- dComb[, .(HT = 2 * mean(f_A) * mean(f_B)), by = pos]   # get HT

SNP_fixed <- dRes[HT == 0, ][["pos"]]  # vector with SNP to remove


## function to get Fst between 2 pop:
get_Fst2pop <- function(f_Pop1, f_Pop2, name_pop1, name_pop2, fixed_allelesTRUE) {
  require(data.table)
  # filtering for globally fixed allels
  dPop1_filt <- f_Pop1[!pos %in% fixed_allelesTRUE, ]
  dPop2_filt <- f_Pop2[!pos %in% fixed_allelesTRUE, ]
  
  dComb <- rbind(dPop1_filt[, region := name_pop1], 
                 dPop2_filt[, region := name_pop2])
  if (nrow(dComb) != nrow(dPop1_filt)*2) {stop("some SNP are missing in one pop")}
  ## get HT, HS and FST
  dRes <- dComb[, .(HT = (2 * mean(f_A) * mean(f_B)), 
                    HS = mean(2 * f_A * f_B)), by = pos] %>%
      .[, Fst := 1 - (HS / HT)]
}

dFst_AF_EA <- get_Fst2pop(f_AF, f_EA, "AF", "EA", SNP_fixed)
dFst_AF_WE <- get_Fst2pop(f_AF, f_WE, "AF", "WE", SNP_fixed)
dFst_EA_WE <- get_Fst2pop(f_AF, f_EA, "EA", "WE", SNP_fixed)

## NOTE: I could get a loop with all possible 2 population combinations (42 in total) to get Fst for each one of them. and then inside that loop also add combined plots from functions that I wrote below.
## in the end I can pool Fst statistics from different population combinations and try to look how those Fst differ between populations?

mean(dFst_AF_EA$Fst)

##########################################################################
## getting regions of Fst.
## function to get 10 regions of 100 SNP with highest Fst values
getXSNPregion <- function(dFst, X_SNP = 100, NRegion = 10, ...) {
  dFst_bin <- dFst[, .(ID = (as.integer(1:nrow(dFst)/X_SNP)), pos, Fst)] %>%
    .[, .(Fst = mean(Fst, na.rm = TRUE),
          start = min(pos), 
          end = max(pos), 
          B = mean(pos, na.rm = TRUE)), by = ID ] %>%
    .[order(Fst, decreasing = TRUE)] %>% 
    head(n = NRegion)
  
}

## function to get 10 regions of 100 KB with highest Fst values
getBinRegion <- function(dFst, BW = 100000, NRegion = 10, ...) {
   dFst_bin <- dFst[, .(B = (as.integer(pos/BW) + 0.5) * BW, Fst)] %>%
    .[, .(Fst = mean(Fst, na.rm = TRUE), start = B - 0.5*BW, 
        end = B + 0.5*BW), by = B ] %>%
    .[order(Fst, decreasing = TRUE)]  %>%
    head(n = NRegion)
}


AF_EA.HFst.SNP <- getXSNPregion(dFst_AF_EA) %>% setkey(start, end)
AF_WE.HFst.SNP <- getXSNPregion(dFst_AF_WE) %>% setkey(start, end)
EA_WE.HFst.SNP <- getXSNPregion(dFst_EA_WE) %>% setkey(start, end)


AF_EA.HFst.KB <- getBinRegion(dFst_AF_EA) %>% setkey(start, end)
AF_WE.HFst.KB <- getBinRegion(dFst_AF_WE) %>% setkey(start, end)
EA_WE.HFst.KB <- getBinRegion(dFst_EA_WE) %>% setkey(start, end)

#########################################################################################
saveRDS(list(AF_EA.HFst.KB, AF_WE.HFst.KB, EA_WE.HFst.KB), file = paste0(output, "FstSelReg_KB.RDS"))
saveRDS(list(AF_EA.HFst.SNP, AF_WE.HFst.SNP, EA_WE.HFst.SNP), file = paste0(output, "FstSelReg_SNP.RDS"))




```


```{r plot_pairwise_fst}


## function to plot Fst by chosen bin (by physical distance)
plot_KBBin_Fst <- function(dFst, bin_width, name_plot, 
                             psize = 0.9, palpha = 1, ...) {
  dFst_bin <- dFst[, .(B = (as.integer(pos/bin_width) + 0.5) * bin_width, Fst)] %>%
  .[, .(Fst = mean(Fst, na.rm = TRUE)), by = B ]
  pBinFst <- dFst_bin %>%
  ggplot(aes(B, Fst)) + 
    theme_bw() +
    xlab("") +
    #geom_point(size = psize, alpha = palpha) +
    geom_line(size = 0.2) +
    ggtitle(name_plot) 
}

## function to plot mean Fst for bins with X SNP in the next to each other (despite distance, just by the order the SNP go on chromosome)
plot_XSNPBin_Fst <- function(dFst, X_SNP_in_bin, name_plot,
                             psize = 0.9, palpha = 1, ...) {
  dFst_bin <- dFst[, .(ID = (as.integer(1:nrow(dFst)/X_SNP_in_bin)), pos, Fst)] %>%
    .[, .(Fst = mean(Fst, na.rm = TRUE), 
          B = mean(pos, na.rm = TRUE)), by = ID ]
  pBinFst <- dFst_bin %>%
  ggplot(aes(B, Fst)) + 
    ggtitle(name_plot) +
    theme_bw() +
    xlab("") +
    geom_point(size = psize, alpha = palpha)
}


SNP_bin <- 100   # 100 SNP in a bin
bin_width10 <- 10000   # 10 kb windows
bin_width100 <- 100000   # 100 kb windows

p100_AF_EA <- plot_KBBin_Fst(dFst_AF_EA, bin_width100, "Africa & East Asia") + 
  geom_rect(data = AF_EA.HFst.KB, aes(xmin = start, xmax = end, ymin = 0, ymax = max(Fst) ), 
            col = "red", size = 0.15, alpha = 0)
p100_AF_WE <- plot_KBBin_Fst(dFst_AF_WE, bin_width100, "Africa & West Eurasia") +
  geom_rect(data = AF_WE.HFst.KB, aes(xmin = start, xmax = end, ymin = 0, ymax = max(Fst)), 
            col = "red", size = 0.15, alpha = 0)
p100_EA_WE <- plot_KBBin_Fst(dFst_EA_WE, bin_width100, "West Eurasia & East Asia") +
  geom_rect(data = EA_WE.HFst.KB, aes(xmin = start, xmax = end, ymin = 0, ymax = max(Fst)), 
            col = "red", size = 0.15, alpha = 0)
grid.arrange(p100_AF_EA, p100_AF_WE, p100_EA_WE)




p10_AF_EA <- plot_KBBin_Fst(dFst_AF_EA, bin_width10, "AF_EA, 10KB windows", 0.8, 0.3) 
p10_AF_WE <- plot_KBBin_Fst(dFst_AF_WE, bin_width10, "AF_WE, 10KB windows", 0.8, 0.3)
p10_EA_WE <- plot_KBBin_Fst(dFst_EA_WE, bin_width10, "EA_WE, 10KB windows", 0.8, 0.3)
# ggsave(grid.arrange(p10_AF_EA, p10_AF_WE, p10_EA_WE), 
#        file = paste0(pic_path, "Fst_10kb_pairwiseComp", ".jpeg"), device = "jpeg",
#        units = "in", width = 9, height = 6)

pSNPbin_AF_EA <- plot_XSNPBin_Fst(dFst_AF_EA, SNP_bin, "AF_EA, ~100 SNP bins", 0.8, 0.5) 
pSNPbin_AF_WE <- plot_XSNPBin_Fst(dFst_AF_WE, SNP_bin, "AF_WE, ~100 SNP bins", 0.8, 0.5)
pSNPbin_EA_WE <- plot_XSNPBin_Fst(dFst_EA_WE, SNP_bin, "EA_WE, ~100 SNP bins", 0.8, 0.5)
# grid.arrange(pSNPbin_AF_EA, pSNPbin_AF_WE, pSNPbin_EA_WE)


tmp <- dFst_AF_EA[Fst > quantile(dFst_AF_EA$Fst, 0.995, na.rm = TRUE)] %>%
    .[, start := pos] %>%
    .[, end := pos]




############################################################################
saveRDS(list(HFst_AF_EA, HFst_AF_WE, HFst_EA_WE), file = paste0(output, "HighFstRegions.RDS"))
ggsave(grid.arrange(p100_AF_EA, p100_AF_WE, p100_EA_WE), 
       file = paste0(pic_path, "Fst_3pop.jpeg"), device = "jpeg", units = "in", height = 6, width = 8.5)
```


```{r LD calc}

## calculate mean and median correlation values for ~100 SNP that are next to each other
## dAF_convMT -- numerix matrix with SNP in columns
calcXSNP_cor <- function(dSnp, dAF_convMT, XSNP = 100, ...) {
  require(data.table)
  nSNP <- dim(dAF_convMT)[[2]]   # number of SNP
  reg <- as.integer((1:nSNP) / XSNP)   # devide SNP to regions
  dReg.cor <- data.frame()   # iniciate result data.frame
  for (i in unique(reg)) {
    print(i)
    res.cor <- cor(dAF_convMT[, reg == i]) 
    #res.cor[lower.tri(res.cor, diag = TRUE)] <- NA   # optional. experiment
    diag(res.cor) <- NA
    dReg.cor <- rbind(dReg.cor, 
                      data.frame(regionID = i, 
                                 pos = mean(dSnp$pos[reg == i], na.rm = TRUE), 
                                 mean.corr = mean((res.cor)^2, na.rm = TRUE) 
                      ))
  }
  return(dReg.cor)
}


dcor_AF <- calcXSNP_cor(dSnp, mAF, 100)
dcor_EA <- calcXSNP_cor(dSnp, mEA, 100)
dcor_WE <- calcXSNP_cor(dSnp, mWE, 100)

pLD_WE <- ggplot(dcor_WE, aes(pos, mean.corr)) +
  #geom_point(size = 0.8) + 
  geom_line(size = 0.2) +
  theme_bw() +
  ggtitle("West Eurasia") +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  xlab("") +
  ylab(expression(r^2))
pLD_AF <- ggplot(dcor_AF, aes(pos, mean.corr)) +
  #geom_point(size = 0.8) + 
  geom_line(size = 0.2) +
  theme_bw() +
  ggtitle("Africa") +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  xlab("") +
  ylab(expression(r^2))
pLD_EA <- ggplot(dcor_EA, aes(pos, mean.corr)) +
  geom_line(size = 0.2) +
  #geom_point(size = 0.8) + 
  theme_bw() +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  ggtitle("East Asia") +
  xlab("") +
  ylab(expression(r^2))

grid.arrange(pLD_AF, pLD_WE, pLD_EA)

####### LD for X KB window ##############################################################
calcSnpBin_cor <- function(dSnp, dAF_conMT, BW = 100000, ...) {
  require(data.table)
  reg <- as.integer((dSnp$pos / BW))
  snp_region_ID <- as.numeric(names(which(table(reg) != 1)))
  dReg.cor <- data.table()   # iniciate result data.frame
  for (i in snp_region_ID) {
    print(i)
    res.cor <- cor(dAF_conMT[, reg == i]) 
    #res.cor[lower.tri(res.cor, diag = TRUE)] <- NA   # optional. experiment
    diag(res.cor) <- NA
    dReg.cor <- rbind(dReg.cor, 
                      data.table(B = i, 
                                 n_snp = nrow(res.cor),
                                 mean.corr = mean(res.cor, na.rm = TRUE))
                      )
  }
  return(dReg.cor)
}

dcor_AF <- calcXSNP_cor(dSnp, mAF)
dcor_EA <- calcXSNP_cor(dSnp, mEA)
dcor_WE <- calcXSNP_cor(dSnp, mWE)

pWE_LD <- ggplot(dcor_WE, aes(pos, mean.corr)) +
  #geom_point(size = 0.8) + 
  geom_line(size = 0.2) +
  theme_bw() +
  ggtitle("West Eurasia") +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  xlab("") +
  ylab(expression(r^2))
pAF_LD <- ggplot(dcor_AF, aes(pos, mean.corr)) +
  #geom_point(size = 0.8) + 
  geom_line(size = 0.2) +
  theme_bw() +
  ggtitle("Africa") +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  xlab("") +
  ylab(expression(r^2))
pEA_LD <- ggplot(dcor_EA, aes(pos, mean.corr)) +
  #geom_point(size = 0.8) + 
  geom_line(size = 0.2) +
  theme_bw() +
  geom_hline(yintercept = median(dcor_WE$mean.corr, na.rm = TRUE), size = 0.5,  linetype = 2, col = "gray") +
  ggtitle("East Asia") +
  xlab("") +
  ylab(expression(r^2))

grid.arrange(pAF_LD, pWE_LD, pEA_LD)
ggsave(grid.arrange(pAF_LD, pWE_LD, pEA_LD), filename = paste0(pic_path, "LD_3pop.jpeg"), 
       device = "jpeg", units = "in", height = 6, width = 8.5)

```




```{r}
## SNP density
temp <- density(dSnp$pos, bw = 10)
jpeg(filename = paste0(pic_path, "SNP_DENSITY.jpeg"))
plot(temp, main = "SNP density by position")
polygon(c(temp$x, rev(temp$x)), 
        c(temp$y, rep(0,length(temp$x))),
col = "blue")
dev.off()


###### Waterson + Fu Li D* ##########################################################

calcWTheta <- function(mAF, dSnp, BW = 100000, ...) {
  nInd <- dim(mAF)[[1]]
  nSNP <- dim(mAF)[[2]]
  
  ## waterson's estimator denomintor
  waterson.dn <- sum(1 / seq_len(nInd - 1))
  
  ## counting haplotype values for each SNP
  n1 <- apply(mAF, 2, function(x) sum(x == 1, na.rm = TRUE))
  n0 <- apply(mAF, 2, function(x) sum(x == 0, na.rm = TRUE))
  nNA <- apply(mAF, 2, function(x) sum(is.na(x)))
  
  ## watersons estimator value for X KB window
  dHapBySnp <- data.table(pos = dSnp$pos, n1 = n1, n0 = n0, nNA = nNA) %>%
    .[, seg := !(n1 + nNA == nInd | n0 + nNA == nInd)] %>%
    .[, B := (as.integer(pos / BW)) * BW ]
  
  dHapBin <- dHapBySnp[, .(nSegS = sum(seg), 
                           nSNP = .N, 
                           median.n1 = median(n1), 
                           median.n0 = median(n0)), by = B]
    
  # adding singleton # in bin for Fu&Li D*
  dHapBin <- merge(dHapBin, dHapBySnp[n0 == 1, .(n_0singl = .N), by = B], by = "B")   
  
  dHapBin[, wTheta := nSegS / waterson.dn] %>%
    .[, FuLiDiff := nSegS - waterson.dn*n_0singl] %>%
    .[, FuLiD := FuLiDiff / sd(FuLiDiff)]

return(dHapBin)
}

EA_wTheta <- calcWTheta(mEA, dSnp)


###### Tajima ################################################################
## Tajima's denominator
calcTTheta <- function(mAF, dSnp, BW = 100000, ...) {
  
  ## getting the denominator
  nInd <- dim(mAF)[[1]]  
  tajima.dn <- (nInd * (nInd - 1)) / 2
  
  ### finding pairwise differences between sequences ####
  # Preparations
  dComb <- cbind(dSnp[, "pos"], t(mAF))   # combining SNP positions and separated genotype
  dComb[, B := (as.integer(pos / BW))*BW] 
  dij <- data.table(B = unique(dComb$B))   
  n <- which(!colnames(dComb) %in% c("pos", "B"))   # col id of samples
  
  ## loop to find pairwise difference
  for (i in n[1:length(n)]) {
   for (j in n[1:length(n)]) {
      if (i < j) {
      sel.chrom <- colnames(dComb)[ c(i, j)]
      dij_name <- paste0("C:", i, "_", j)
      print(dij_name)
      # print(sel.chrom)
      dTmp <- dComb[, c(sel.chrom, "B"), with = FALSE] 
      colnames(dTmp) <- c("X", "Y", "B")
      dRes <- dTmp[, sum(X != Y, na.rm = TRUE), by = B]
      colnames(dRes) <- c("B", dij_name)
      dij <- merge(dij, dRes, by = "B")
      } 
   }
  }
  
  stopifnot(dim(dij[, -c("B")])[[2]] == tajima.dn)
  dPwDiff <- data.table(B = dij$B, 
                        pwd = apply(dij[, -c("B")], 1, sum)) %>%
    .[, tTheta := pwd / tajima.dn]
}

calcTD <- function(mAF, dSnp, BW = 100000, ...) {
  print("calculating WTheta")
  wTheta <- calcWTheta(mAF, dSnp)
  print("calculating TTheta")
  tTheta <- calcTTheta(mAF, dSnp)
  
  dDivStat <- merge(wTheta[, c("B", "nSegS", "n_0singl", "FuLiD", "wTheta")], 
                    tTheta[, c("B", "tTheta")], by = "B") %>%
    .[, Tdiff := tTheta - wTheta] %>%
    .[, TD := Tdiff / sd(Tdiff)] 
  dDivStat$Tdiff <- NULL
  return(dDivStat)
}
  
TD_AF <- calcTD(mAF, dSnp)  
TD_WE <- calcTD(mWE, dSnp)
TD_EA <- calcTD(mEA, dSnp)

pAF.TD <- ggplot(TD_AF, aes(B, TD)) + 
  geom_line(size = 0.3) +
  geom_point(data = TD_AF[TD > 0], aes(B, TD), col = "red", size = 0.4) +
  xlab("") +
  theme_bw() +
  ylab("Tajima D") +
  ggtitle("Africa")
pWE.TD <- ggplot(TD_WE, aes(B, TD)) + 
  geom_line(size = 0.3) +
  geom_point(data = TD_WE[TD > 0], aes(B, TD), col = "red", size = 0.4) +
  xlab("") +
  theme_bw() +
  ylab("Tajima D") +
  ggtitle("West Eurasia")
pEA.TD <- ggplot(TD_EA, aes(B, TD)) + 
  geom_point(data = TD_EA[TD > 0], aes(B, TD), col = "red", size = 0.4) +
  geom_line(size = 0.3) +
  xlab("") +
  theme_bw() +
  ylab("Tajima D") +
  ggtitle("East Asia")

grid.arrange(pAF.TD, pWE.TD, pEA.TD)


pAF.WT <- ggplot(TD_AF, aes(B, wTheta)) + 
  geom_line(size = 0.3) +
  xlab("") +
  theme_bw() +
  ylab(expression(theta["waterson"])) +
  ggtitle("Africa")
pWE.WT <- ggplot(TD_WE, aes(B, wTheta)) + 
  geom_line(size = 0.3) +
  xlab("") +
  theme_bw() +
  ylab(expression(theta["waterson"])) +
  ggtitle("West Eurasia")
pEA.WT <- ggplot(TD_EA, aes(B, wTheta)) + 
  geom_line(size = 0.3) +
  xlab("") +
  theme_bw() +
  ylab(expression(theta["waterson"])) +
  ggtitle("East Asia")

grid.arrange(pAF.WT, pWE.WT, pEA.WT)


ggsave(grid.arrange(pAF.WT, pWE.WT, pEA.WT), filename = paste0(pic_path, "WT_3pop.jpeg"), 
       device = "jpeg", units = "in", height = 6, width = 8.5)

ggsave(grid.arrange(pAF.TD, pWE.TD, pEA.TD), filename = paste0(pic_path, "TD_3pop.jpeg"), 
       device = "jpeg", units = "in", height = 6, width = 8.5)

```

