---
title: "Region_overlap"
author: "Vita"
date: "May 21, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(ggplot2)
library(dplyr)
library(knitr)
```


```{r}
## paths
input <- "../input/"
output <- "../output/"
pic_path <- "../graphs/"

## gene annotation
gtf <- rtracklayer::import(paste0(input, "gencode.v17.annotation.gtf")) %>% 
  as.data.table() %>%    # convert region data.frame
  .[seqnames == "chrX", ] %>%    # to select only X chromosome
  setkey(start, end)

## significant sites

## Fst
tmp <- readRDS(file = paste0(output, "FstSelReg_KB.RDS"))
#saveRDS(list(AF_EA.HFst.SNP, AF_WE.HFst.SNP, EA_WE.HFst.SNP), file = paste0(output, "FstSelReg_SNP.RDS")
AF_EA.HFst.KB <- tmp[[1]]
AF_WE.HFst.KB <- tmp[[2]]
EA_WE.HFst.KB <- tmp[[3]]

## IHS
tmp <- readRDS(file = paste0(output, "rehh.selectedSNP.RDS"))
AF.ihs.selSNP <- tmp[[1]]
EA.ihs.selSNP <- tmp[[2]]
WE.ihs.selSNP <- tmp[[3]]
## XPEHH
AF_EA.ehh.selSNP <- tmp[[4]]
AF_WE.ehh.selSNP <- tmp[[5]]
EA_WE.ehh.selSNP <- tmp[[6]]


```


```{r}

### 100 KB Fst windows and genes overlap
AF_EA_F <- foverlaps(AF_EA.HFst.KB, gtf, type = "any", nomatch = 0) %>% 
  .[type == "gene", .(gene_name,  gene_type)] %>% 
  unique.data.frame() %>% as.data.table()
AF_WE_F <- foverlaps(AF_WE.HFst.KB, gtf, type = "any", nomatch = 0) %>% 
  .[type == "gene", .(gene_name,  gene_type)] %>% 
  unique.data.frame() %>% as.data.table()
EA_WE_F <- foverlaps(EA_WE.HFst.KB, gtf, type = "any", nomatch = 0) %>%
  .[type == "gene", .(gene_name, gene_type)] %>% 
  unique.data.frame() %>% as.data.table()
F_tmp <- rbind(EA_WE_F[, comparison := "EA_WE"], 
               AF_EA_F[, comparison := "AF_EA"], 
               AF_WE_F[, comparison := "AF_WE"])
FstComp <- dcast(F_tmp, gene_name + gene_type ~ comparison) %>%
  .[order(EA_WE), c(1, 3:5, 2)]
FstComp <- apply(FstComp, 1, function(x) {
  gsub("EA_WE|AF_EA|AF_WE", "+", x)
})
knitr::kable(t(FstComp), format = "latex")


# ### 100 SNP Fst windows
# AF_EA_F <- foverlaps(AF_EA.HFst.SNP, gtf, type = "any", nomatch = 0) %>% 
#   .[type == "gene", .(gene_name,  gene_type)] %>% 
#   unique.data.frame() %>% as.data.table()
# AF_WE_F <- foverlaps(AF_WE.HFst.SNP, gtf, type = "any", nomatch = 0) %>% 
#   .[type == "gene", .(gene_name,  gene_type)] %>% 
#   unique.data.frame() %>% as.data.table()
# EA_WE_F <- foverlaps(EA_WE.HFst.SNP, gtf, type = "any", nomatch = 0) %>%
#   .[type == "gene", .(gene_name, gene_type)] %>% 
#   unique.data.frame() %>% as.data.table()
# F_tmp <- rbind(EA_WE_F[, comparison := "EA_WE"], 
#                AF_EA_F[, comparison := "AF_EA"], 
#                AF_WE_F[, comparison := "AF_WE"])
# FstComp <- dcast(F_tmp, gene_name + gene_type ~ comparison) %>%
#   .[order(EA_WE), c(1, 3:5, 2)]
# FstComp <- apply(FstComp, 1, function(x) {
#   gsub("EA_WE|AF_EA|AF_WE", "+", x)
# })
# knitr::kable(t(FstComp), format = "latex")

############################################################
# IHS and genes overlap
AF.ihsG <- foverlaps(AF.ihs.selSNP, gtf, type = "any", nomatch = 0) %>% 
  .[type == "gene", .(gene_name, gene_type)] %>%
    unique.data.frame() %>% as.data.table()
EA.ihsG <- foverlaps(EA.ihs.selSNP, gtf, type = "any", nomatch = 0) %>% 
  .[type == "gene", .(gene_name, gene_type)] %>%
    unique.data.frame() %>% as.data.table()
WE.ihsG <- foverlaps(WE.ihs.selSNP, gtf, type = "any", nomatch = 0) %>% 
  .[type == "gene", .(gene_name, gene_type)] %>%
    unique.data.frame() %>% as.data.table()

ihs_tmp <- rbind(AF.ihsG[, comparison := "AF"], 
                 EA.ihsG[, comparison := "EA"], 
                 WE.ihsG[, comparison := "WE"])
ihsComp <- dcast(ihs_tmp, gene_name + gene_type ~ comparison) %>%
  .[order(AF), c(1, 3:5, 2)]
ihsComp <- apply(ihsComp, 1, function(x) {
  gsub("EA|AF|WE", "+", x)
}) %>% t()

ihsComp
knitr::kable(ihsComp, format = "latex")

############################################################
# XPEHH and genes overlap
  AF_EA.EHHG <- foverlaps(AF_EA.ehh.selSNP, gtf, type = "any", nomatch = 0) %>% 
    .[type == "gene", .(gene_name, gene_type)] %>%
      unique.data.frame() %>% as.data.table()
  AF_EA.EHHG <- foverlaps(AF_WE.ehh.selSNP, gtf, type = "any", nomatch = 0) %>% 
    .[type == "gene", .(gene_name, gene_type)] %>%
      unique.data.frame() %>% as.data.table()
  EA_WE.EHHG <- foverlaps(EA_WE.ehh.selSNP, gtf, type = "any", nomatch = 0) %>% 
    .[type == "gene", .(gene_name, gene_type)] %>%
      unique.data.frame() %>% as.data.table()
  
  ehh_tmp <- rbind(AF_EA.EHHG[, comparison := "AF_EA"], 
                   AF_EA.EHHG[, comparison := "AF_WE"], 
                   EA_WE.EHHG[, comparison := "EA_WE"])
  ehh_tmp[, .N, comparison]
 #  ehh_Comp <- dcast(ehh_tmp, gene_name + gene_type ~ comparison) %>%
 #    .[order(AF_EA), c(1, 3:5, 2)]
 #  ehh_Comp <- apply(ehh_Comp, 1, function(x) {
 #    gsub(paste0(c( "AF_EA", "AF_WE", "EA_WE"), collapse = "|"), "+", x)
 #  }) %>% t()
  ehh_Comp <- dcast(ehh_tmp, gene_name + gene_type ~ comparison) %>%
    .[order(AF_WE), c(1, 3:4, 2)]
  ehh_Comp <- apply(ehh_Comp, 1, function(x) {
    x <- gsub("0", "NA", x)
    gsub("2|1", "+", x)
  }) %>% t()

ehh_Comp
knitr::kable(ehh_Comp, format = "latex")
```

