---
title: "Fst_try"
author: "Vita"
date: "May 2, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(data.table)
library(ggplot2)
```


```{r import_data}
## paths
input <- "../input/"
output <- "../output/"
pic_path <- "../graphs/"

## import genotype data
dAF <- fread(paste0(input, "genotypes_AF"), header = FALSE)
dEA <- fread(paste0(input, "genotypes_EA"), header = FALSE)
dSA <- fread(paste0(input, "genotypes_SA"), header = FALSE)
dWE <- fread(paste0(input, "genotypes_WE"), header = FALSE)
dAM <- fread(paste0(input, "genotypes_AM"), header = FALSE)
dO <- fread(paste0(input, "genotypes_O"), header = FALSE)
dCAS <- fread(paste0(input, "genotypes_CAS"), header = FALSE)

## Snp data
dSnp <- fread(paste0(input, "snps_filtered")) 
colnames(dSnp) <- c("ID", "chrom", "pos", "base", "alt")
```


```{r removing missing data}
##### how many snp there are that completely miss data? ###################
## function to get that
get_NA_perc_forSNP <- function(Data, ...) {
  nNA_snp <- apply(Data, 1, function(x) sum(is.na(x)))
  max_nHap <- dim(Data)[2]
  df_NA_stat <- data.table(ID = 1:length(nNA_snp), 
                           n_na = nNA_snp,
                           perc_NA = nNA_snp / max_nHap) 
  #bin_width <- 100   # 100 SNP
  #df_NA_stat[, plot_bin := (as.integer(ID / bin_width) + 0.5) * bin_width]  
  return(df_NA_stat)
}

df_SnpNa_AF <- get_NA_perc_forSNP(dAF)
df_SnpNa_EA <- get_NA_perc_forSNP(dEA)
df_SnpNa_SA <- get_NA_perc_forSNP(dSA)
df_SnpNa_WE <- get_NA_perc_forSNP(dWE)
df_SnpNa_AM <- get_NA_perc_forSNP(dAM)
df_SnpNa_O <- get_NA_perc_forSNP(dO)
df_SnpNa_CAS <- get_NA_perc_forSNP(dCAS)


table(df_SnpNa_AF$n_na)   # 113505 completely not called
table(df_SnpNa_EA$n_na)   # 113505 completely not called
table(df_SnpNa_SA$n_na)   # 113505 completely not called
table(df_SnpNa_WE$n_na)   # all good
table(df_SnpNa_O$n_na)   # all good
table(df_SnpNa_CAS$n_na)    # all good
table(df_SnpNa_AM$n_na)   # all good

p <- ggplot(df_SnpNa_AF, aes(ID, perc_NA)) + 
      geom_point() +
      xlab("100 SNP bins") +
      ylab("mean percentage of SNP missing data")
ggsave(p, file = paste0(pic_path, "AF_missing_SNP", ".jpeg"), device = "jpeg")
## it looks like AF, EA and SA have completely uncalled end of chromosome.
## other pop are ok.
## will exclude those SNP from all populations to make it easier

## checking if the same SNP are uncalled in AF, SA and EA pop
table(df_SnpNa_AF$perc_NA == 1, df_SnpNa_SA$perc_NA == 1)   # ok
table(df_SnpNa_AF$perc_NA == 1, df_SnpNa_EA$perc_NA == 1)   # ok, so I can use the same for all of them 

removeNASnp_SaveNewData <- function(Data, snp_keep, save_name, ..) {
  write.table(Data[snp_keep, ], file = paste0(output, save_name, "_removeNAsnp"), 
              row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
}
snp_keep <- df_SnpNa_AF$perc_NA != 1
removeNASnp_SaveNewData(dAF, snp_keep, "dAF")
removeNASnp_SaveNewData(dEA, snp_keep, "dEA")
removeNASnp_SaveNewData(dSA, snp_keep, "dSA")
removeNASnp_SaveNewData(dWE, snp_keep, "dWE")
removeNASnp_SaveNewData(dO, snp_keep, "dO")
removeNASnp_SaveNewData(dCAS, snp_keep, "dCAS")
removeNASnp_SaveNewData(dAM, snp_keep, "dAM")
removeNASnp_SaveNewData(dSnp, snp_keep, "dSnpPos")
```












