---
title: "intersect with gene annotation"
author: "Vita"
date: "May 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(ggplot2)
library(dplyr)
library(GenomicRanges)

```

```{r}
## paths
input <- "../input/"
output <- "../output/"
pic_path <- "../graphs/"

## significant sites
dSigW <- fread(paste0(output, "dSigWithin_ihs.txt"))


## gene annotation
require(GenomicRanges)
## genomic ranges way
gtf <- rtracklayer::import(paste0(input, "gencode.v17.annotation.gtf"))
gtf <- gtf[seqnames(gtf) == 'chrX']   # to select only X chromosome
## convert region data.frame
tmp <- with(dSigW, GRanges(seqnames = rep("chrX", nrow(dSigW)), 
                             IRanges(start = start, end = end)))

hits <- findOverlaps(gtf, tmp)  
ranges(gtf)[queryHits(hits)]
res <- gtf[queryHits(hits), ] %>% as.data.frame()
unique(res$gene_name)

## data.table way
require(data.table)
gtf_df <- as.data.table(gtf)   # used gft from before. nicer format
gtf_df <- gtf_df[seqnames == "chrX", ]   # to select X chr
setkey(gtf_df, seqnames, start, end)   # need to index

setkey(dSigW, seqnames, start, end)
dRes <- foverlaps(dSigW, gtf_df, type = "any", nomatch = 0)
dRes[type == "gene", ]


```



```{r overlaping with gene annotation}

## interesting... 3 genes are only match the SNP


```


```{r}
class(gtf)
example = GRanges(seqnames=c(rep("chr1", 4), rep("chr2", 4)),
	                  ranges = IRanges(start = c(10, 32, 59, 79, 11, 22, 23, 41),
	                                   end = c(42, 51, 76, 89, 12, 31, 46, 49)), 
	                  strand = rep(c("+", "-"), 4), 
	                  seqlengths = c(chr1 = 120, chr2 = 70) )
gr <- GRanges(seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
              ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
              strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
              score = 1:10,
              GC = seq(1, 0, length = 10))
ten_sig_reg <- dSigW[1:10]
ten_sig_reg$POSITION <- dSigPos$start[1:10 ]
tmp = GRanges(seqnames = c(rep("chrX", nrow(ten_sig_reg)),
	                ranges = IRanges(start = ten_sig_reg$POSITION,
	                                   end = ten_sig_reg$POSITION)))

tmp <- with(ten_sig_reg, GRanges(seqnames, IRanges(start = POSITION, end = POSITION)))
tmp <- with(dSigPos, GRanges(seqnames = rep("chrX", nrow(dSigPos)), 
                             IRanges(start = start, end = end)))

hits <- findOverlaps(gtf, tmp)  
ranges(gtf)[queryHits(hits)]
res <- gtf[queryHits(hits), ] %>% as.data.frame()
unique(res$gene_name)



filter(res, type == "gene")
hits2 <- as.data.frame(hits)
head(hits2)
dSigPos[2, ]
gtf[c(1277, 1278, 1297), ]
dRes <- gtf[subjectHits(hits)] %>%
    as.data.frame()
filter(dRes, type == "transcript")
```


```{r}
install.packages("LDheatmap")
library(LDheatmap)
data(CEUData)
```



```{r}
dAF
snpgdsCreateGeno("test.gds", genmat = hapmap_geno$genotype,
    sample.id = hapmap_geno$sample.id, snp.id = hapmap_geno$snp.id,
    snp.chromosome = hapmap_geno$snp.chromosome,
    snp.position = hapmap_geno$snp.position,
    snp.allele = hapmap_geno$snp.allele, snpfirstdim=TRUE)
```

